File: binaryTree_iterator.java
// Time Complexity : next: amortised O(1), hasNext: O(1)
// Space Complexity : next: O(N), hasNext: O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Couldn't solve by own


// Your code here along with comments explaining your approach:
//Since this is a binary search tree, the lowest elemet is present at the end of leftmost leafnode. Therefore, first iterate through the leftside nodes in iterator.
//Then, in next pop the topmost element. Then, check whether a rightside element is present. Then insert all of it's left side elements.

class BSTIterator {

    //initialize our stack
    Stack<TreeNode> s;
    
    public BSTIterator(TreeNode root) {
        s = new Stack<>();
        
        TreeNode cursor = root;
        
        //pushing left side elements into stack
        while(cursor != null){
            s.push(cursor);
            cursor = cursor.left;
        }
    }
    
    /** @return the next smallest number */
    public int next() {
        //node at top of stack is smallest element
        TreeNode res = s.pop();
        
        //there can nodes present on right side
        TreeNode cursor = res.right;
        
        //push all it's left nodes to stack
        while(cursor != null){
            s.push(cursor);
            cursor = cursor.left;
        }
        
        return res.val;
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !s.isEmpty();
    }
}

File: binaryTreeIterator.py
'''
time complexity: O(1)
space complexity: O(h)
h is height of tree
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self.root = root
        self.dfs(root)
        
        

    def next(self) -> int:
        temp = self.stack.pop()
        self.dfs(temp.right)
        return temp.val

    def hasNext(self) -> bool:
        if(len(self.stack)!=0):
            return True
        return False
    def dfs(self,node):
        while(node!=None):
            self.stack.append(node)
            node=node.left


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()